5.0 Draft/IDD/IDDs Orchestration System/eu.arrowhead.orchestration-pull-http-json.yml

```diff
@@ -0,0 +1,303 @@
+        on a existing SoS configuration or Orchestration rules. 
+
+        Information of the requesting system is mandatory. Requested service is optional. If a requested service object is specified,
+        the service definition and at least one interface must be specified. 
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ServiceRequestForm'
+      responses:
+        '200':
+          description: Successful orchestration response.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/OrchestrationResponse'
+
+  /orchestration/orchestration-pull/{id}:
+    get:
+      summary: Get Orchestration by system id
+      parameters:
+        - name: id
+          in: path
+          required: true
+          description: ID of requesting system
+          schema:
+            type: string
+      responses:
+        '200':
+          description: Successful response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/OrchestrationResponse'
+        '404':
+          description: System id not found
+          content:
+            application/json:
+              example:
+                error: Orchestration failed, system id not found. 
+
+
+components:
+  schemas:
+    Address:
+      description: |
+        Identifies a _transport_ and a _location_.
+
+        The _transport_ identifies the base protocol that facilitates addressing a specific service
+        instance. Examples of transports that can be supported are `tcp4`, `tcp6`, `udp4`, `udp6`
+        and `unix`. A protocol only counts as a transport if it both (A) provides a way of
+        addressing and, by extension, sending messages to service providers and consumers, as well
+        as (B) does not build upon another protocol also providing this capability. I other words,
+        TLS and DTLS are _not_ transports, because they build upon the TCP and UDP protocols, which
+        satisfy condition A.
+
+        What the _location_ part consists of depends on what transport is identified. If the
+        transport is `tcp4` or `udp4`, the location is an IPv4 address expressed as four decimal
+        numbers separated by dots, a colon and a port number, such as in `192.168.3.22:64075`. If
+        the transport is `tcp6` or `udp6`, the location is an IPv6 address, rendered as described in
+        [RFC 5952](https://www.rfc-editor.org/rfc/rfc5952), within square brackets, followed by a
+        colon and a port number. If the transport is `unix`, the location is an absolute filesystem
+        path to a Unix socket file.
+      type: string
+      pattern: ^(?<transport>[^:]+):\w*(?<location>.*)$
+      example: tcp4:192.168.0.7:45326
+
+    Auth-tokens:
```

Comment:
I guess you just took the Arrowhead 4 something IDD and converted into an OpenAPI-specification?

My impression was that most people at Sinetiq are opposed to the idea of distributing authorization tokens through the service orchestration service. @PerOlofsson-Sinetiq Am I wrong?

Reply:
The intention of having auth-tokens is to ensure that the system requesting orchestration is authenticated, but now I see that I put the auth-token in the response rather than in the requestForm. 
I will put the auth-token in the ServiceRequestForm, and remove it from the orchestrationResponse.  
Let me know what you think about it.

Reply:
Do we want the Service Orchestration service to be (A) a facade that helps you use multiple other services, or do we want it to be (B) a service in its own right? The Arrowhead 4 Orchestrator was a facade, but @PerOlofsson-Sinetiq, myself and some others (if my memory is not incorrect) think it should be a service in its own right.

The reason for us wanting this is that it limits the Service Orchestration service to a single objective: _to recommend to a consumer what services it should consume_. If it also is a facade, it has two objectives, where the second one is to help you make use of the functionality of the Service Discovery and Authorization services.

There is nothing wrong with the facade approach. The problem with it is that none of the services it is a facade over are stable. They will change a lot until the project settle on their designs. This means that, until the designs start to settle, the facade becomes a design choke-point. Every change to the Service Discovery and Authorization services may likely require a change to the Service Orchestration service.

To make it not a facade, it needs to simply provide you with the names of the services it thinks you should consume. You (the consumer) are still responsible for looking up the addresses associated with those names via the Service Discovery service. Per and I don't believe authorization tokens should be part of the 5.0 release of Arrowhead, which means that no such tokens need to be distributed at this point.

Reply:
To have this clear, we want to stay away from the "Query facade" that the orchestrator had in the previous implementation. This means that the consuming system no longer asks the orchestrator for "I want x service, with x interface, x security type, etc, etc" but rather "What service should I consume?"? Do we want to make it just like that? If we start adding filtering capabilities to the "recommended services" were do we draw the line?

Reply:
If you ask a friend about recommendations for what ice cream to try, you might add that you are allergic to peanuts, don't like sorbet and usually enjoy dark chocolate. It is OK to add preferences, but they should not be required and usually don't oblige the receiver to take them into account.

Reply:
My issue is then... where do we want to store that "filter information"? If my friend needs to ask the ice cream dude whether certain flavour contains peanuts or not, then I could have made the same question myself, instead of triangulating. The orchestration would need to know all the information about the services to make the filtered recommendations worth it, but that may mean then asking the service-registry whenever a recommendation request comes, which again defeats the purpose. The friend imo should probably be like "your mother told me that you can have either chocolate or vanilla, make your choice". I would say, if we can have the orchestration rules contain basic information such as interface or version, then it makes sense to have it as optional filters.